### DISCUSSION TITLE
`re.compile` receiving generically typed argument gives error

### DISCUSSION THREAD

**Discussion by pvandyken** [microsoft/pyright#5793] [Q&A]
I'm trying to understand why pyright reports an error on the following code:

```python
from __future__ import annotations
import re
from typing import AnyStr

def compile(a: AnyStr | re.Pattern[AnyStr]):
    re.compile(a)
	#		   ~
    # Argument of type "AnyStr@compile | Pattern[AnyStr@compile]" cannot be assigned to parameter "pattern" of type "Pattern[AnyStr@compile]" in function "compile"
	#  Type "str* | bytes* | Pattern[AnyStr@compile]" cannot be assigned to type "Pattern[AnyStr@compile]"
	#    "bytes*" is incompatible with "Pattern[AnyStr@compile]"
```

Since `re.compile` should take either a `AnyStr` or `Pattern` and return a `Pattern` of the same generic type, I wouldn't expect this to have any problem. Does this actually represent unsafe code, or is this a technical limitation in pyright?

**Comment by pvandyken** [comment]
Note that I can clear the error with the following:

```py
from __future__ import annotations
import re
from typing import AnyStr

def compile(a: AnyStr | re.Pattern[AnyStr]):
    _foo: re.Pattern[AnyStr] = re.compile(a)
```

But I don't understand why the version without a specific type annotation can't be inferred

**Reply by erictraut** [reply]
This should type check without errors. I'll create a bug report out of this issue and investigate further.
